### PR TITLE
Add state.reward processing

### PR DESCRIPTION
**Pull request by rodiazet**

### CODE REVIEW

File: test/t8n/t8n.cpp [hunk 1]
```diff
@@ -54,6 +55,13 @@ int main(int argc, const char* argv[])
             output_result_file = argv[i];
         else if (arg == "--output.alloc" && ++i < argc)
             output_alloc_file = argv[i];
+        else if (arg == "--state.reward" && ++i < argc)
+        {
+            if (std::string(argv[i]) == "-1")
+                state_reward = std::numeric_limits<intx::uint256>::max();
```
Comment (axic) [inline test/t8n/t8n.cpp:61]: This is for disabling it? Why not use `std::optional<intx::uint256>` in that case?

File: test/t8n/t8n.cpp [hunk 2]
```diff
@@ -128,6 +136,16 @@ int main(int argc, const char* argv[])
                 }
             }
 
+            if (state_reward != 0 && state_reward != std::numeric_limits<intx::uint256>::max())
+                state.get_or_insert(block.coinbase).balance += state_reward;
+            else if (state_reward == 0)
+                state.touch(block.coinbase);
+            else if (state_reward == std::numeric_limits<intx::uint256>::max())
+            {
+                if (state.get_or_insert(block.coinbase).is_empty())
```
Comment (axic) [inline test/t8n/t8n.cpp:145]: Hm, cleaning up empty accounts is just an EIP-161 rule. Don't we have that implemented?

File: test/t8n/t8n.cpp (hunk 1 above)
Comment (chfast) [reply test/t8n/t8n.cpp:61]: I agree here.

File: test/t8n/t8n.cpp [hunk 3]
```diff
@@ -30,6 +30,7 @@ int main(int argc, const char* argv[])
     fs::path output_dir;
     fs::path output_result_file;
     fs::path output_alloc_file;
+    intx::uint256 state_reward;
```
Comment (chfast) [inline test/t8n/t8n.cpp:33]: Rename it to block reward.

File: test/t8n/t8n.cpp (hunk 2 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:145]: For `<= EIP158` coinbase account is touched even it tx is rejected. retesteth passing -1 as a stare.reward requires zero state to be removed.

File: test/t8n/t8n.cpp [hunk 4]
```diff
@@ -54,6 +55,13 @@ int main(int argc, const char* argv[])
             output_result_file = argv[i];
         else if (arg == "--output.alloc" && ++i < argc)
             output_alloc_file = argv[i];
+        else if (arg == "--state.reward" && ++i < argc)
+        {
+            if (std::string(argv[i]) == "-1")
```
Comment (chfast) [inline test/t8n/t8n.cpp:60]:
```suggestion
            if (std::string_view value{argv[i]}; value == "-1")
```

File: test/t8n/t8n.cpp [hunk 5]
```diff
@@ -54,6 +55,13 @@ int main(int argc, const char* argv[])
...
+                block_reward = {};
+            else
+                block_reward = intx::from_string<intx::uint256>(argv[i]);
```
Comment (chfast) [inline test/t8n/t8n.cpp:63]:
```suggestion
                block_reward = intx::from_string<intx::uint256>(value);
```

File: test/t8n/t8n.cpp [hunk 6]
```diff
@@ -128,6 +136,16 @@ int main(int argc, const char* argv[])
                 }
             }
 
+            if (block_reward.has_value())
+            {
+                if (*block_reward == 0)
+                    state.touch(block.coinbase);
```
Comment (chfast) [inline test/t8n/t8n.cpp:142]: This should handle both cases.

```suggestion
                    state.touch(block.coinbase).balance += *block_reward;
```

File: test/t8n/t8n.cpp [hunk 7]
```diff
@@ -128,6 +136,16 @@ int main(int argc, const char* argv[])
...
+                else
+                    state.get_or_insert(block.coinbase).balance += *block_reward;
+            }
+            else if (rev <= EVMC_TANGERINE_WHISTLE)
```
Comment (chfast) [inline test/t8n/t8n.cpp:146]: Don't we do the same in transaction execution?

File: test/t8n/t8n.cpp (hunk 6 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:142]: Adding 0 does not change so much :)

File: test/t8n/t8n.cpp (hunk 7 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:146]: no, we are touching there here we are erasing.

File: test/t8n/t8n.cpp [hunk 8]
```diff
@@ -54,6 +55,13 @@ int main(int argc, const char* argv[])
             output_result_file = argv[i];
         else if (arg == "--output.alloc" && ++i < argc)
             output_alloc_file = argv[i];
+        else if (arg == "--state.reward" && ++i < argc)
+        {
+            if (std::string_view(argv[i]) == "-1")
+                block_reward = {};
```
Comment (axic) [inline test/t8n/t8n.cpp:61]: Do we actually need to set this here? Can we just have `!= "-1"` case for setting the value?

File: test/t8n/t8n.cpp [hunk 9]
```diff
@@ -128,6 +132,11 @@ int main(int argc, const char* argv[])
                 }
             }
 
+            if (block_reward.has_value())
+                state.touch(block.coinbase).balance += *block_reward;
+            else if (rev <= EVMC_TANGERINE_WHISTLE)  // This behaviour is required by retesteth
+                state.get_accounts().erase(block.coinbase);
```
Comment (axic) [inline test/t8n/t8n.cpp:138]: Hm, this will remove it even when other transactions touched this account. Is that desired?

Comment (rodiazet) [reply test/t8n/t8n.cpp:139]: Sorry. Fore some reason I remove also checking if it's empty. Fixed now

File: test/t8n/t8n.cpp [hunk 10]
```diff
@@ -54,6 +55,9 @@ int main(int argc, const char* argv[])
             output_result_file = argv[i];
         else if (arg == "--output.alloc" && ++i < argc)
             output_alloc_file = argv[i];
+        else if (arg == "--state.reward" && ++i < argc)
+            if (std::string_view(argv[i]) != "-1")
```
Comment (chfast) [inline test/t8n/t8n.cpp:59]: You have to combine all 3 conditions. You can be fancy by doing `argv[i] != "-1"sv`.

File: test/t8n/t8n.cpp [hunk 11]
```diff
@@ -128,6 +132,12 @@ int main(int argc, const char* argv[])
                 }
             }
 
+            if (block_reward.has_value())
+                state.touch(block.coinbase).balance += *block_reward;
```
Comment (chfast) [inline test/t8n/t8n.cpp:136]: This should be done _before_ any transaction execution.

File: test/t8n/t8n.cpp (hunk 9 above)
Comment (chfast) [reply test/t8n/t8n.cpp:139]: I'm still not convinced by this quirk. It looks it replicates a bug in some client implementation. Can you provide a list of failing tests?

File: test/t8n/t8n.cpp (hunk 10 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:59]: error: no matching literal operator for call to 'operator""sv' with arguments of types 'const char *' and 'unsigned long', and no matching literal operator template
        else if (arg == "--state.reward" && ++i < argc && argv[i] != "-1"sv)

Comment (chfast) [reply test/t8n/t8n.cpp:59]: `using namespace std::literals`.

File: test/t8n/t8n.cpp (hunk 11 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:136]: I removed it. It looks that now previously failing test passed. I will check why and let know.

Comment (rodiazet) [reply test/t8n/t8n.cpp:136]: Are you sure? Moving block reward to coinbase before transaction made a lot of test started failing.

Comment (chfast) [reply test/t8n/t8n.cpp:136]: Should be _after_

**Review by chfast — APPROVED** [review]

**Pull request merged**